drivers/can/can_mcp25xxfd.c

```diff
@@ -0,0 +1,1026 @@
+	int ret;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+	ret = mcp25xxfd_readb(dev, MCP25XXFD_REG_CON + 2, &con.byte[2]);
+	k_mutex_unlock(&dev_data->mutex);
+	*mode = con.OPMOD;
+	return ret;
+}
+
+static int mcp25xxfd_set_raw_mode(const struct device *dev, uint8_t mode)
+{
+	struct mcp25xxfd_data *dev_data = DEV_DATA(dev);
+	union mcp25xxfd_con con;
+	int ret;
+
+	while (true) {
+		k_mutex_lock(&dev_data->mutex, K_FOREVER);
+		ret = mcp25xxfd_readw(dev, MCP25XXFD_REG_CON, &con);
+		if (ret < 0 || con.OPMOD == mode) {
+			k_mutex_unlock(&dev_data->mutex);
+			break;
+		}
+
+		if (con.OPMOD == MCP25XXFD_OPMODE_CONFIGURATION) {
+			/* Configuration mode can be switched to any other mode */
+			con.REQMOD = mode;
+		} else if (con.OPMOD == MCP25XXFD_OPMODE_NORMAL_CANFD ||
+			   con.OPMOD == MCP25XXFD_OPMODE_NORMAL_CAN2) {
+			/* Normal modes can only be directly switched to Sleep, Restricted, or Listen-Only modes */
+			if (mode == MCP25XXFD_OPMODE_SLEEP ||
+			    mode == MCP25XXFD_OPMODE_RESTRICTED ||
+			    mode == MCP25XXFD_OPMODE_LISTEN_ONLY) {
+				con.REQMOD = mode;
+			} else {
+				con.REQMOD = MCP25XXFD_OPMODE_CONFIGURATION;
+			}
+		} else if (con.OPMOD == MCP25XXFD_OPMODE_LISTEN_ONLY) {
+			/* Listen-Only mode can be directly switched back to normal modes */
+			if (mode == MCP25XXFD_OPMODE_NORMAL_CANFD ||
+			    mode == MCP25XXFD_OPMODE_NORMAL_CAN2) {
+				con.REQMOD = mode;
+			} else {
+				con.REQMOD = MCP25XXFD_OPMODE_CONFIGURATION;
+			}
+		} else {
+			/* Otherwise, the device must be put into configuration mode first */
+			con.REQMOD = MCP25XXFD_OPMODE_CONFIGURATION;
+		}
+
+		LOG_DBG("OPMOD: #%d, REQMOD #%d", con.OPMOD, con.REQMOD);
+		ret = mcp25xxfd_writeb(dev, MCP25XXFD_REG_CON + 3,
+				       &con.byte[3]);
+		k_mutex_unlock(&dev_data->mutex);
+		if (ret < 0) {
+			break;
+		}
+
+		ret = k_sem_take(&dev_data->mode_sem, K_MSEC(2));
+		if (ret == -EAGAIN) {
+			return CAN_TIMEOUT;
+		}
+	}
+	return ret;
+}
+
+static int mcp25xxfd_set_mode(const struct device *dev, enum can_mode mode)
+{
+	switch (mode) {
+	case CAN_NORMAL_MODE:
+#if defined(CONFIG_CAN_FD_MODE)
+		return mcp25xxfd_set_raw_mode(dev,
+					      MCP25XXFD_OPMODE_NORMAL_CANFD);
+#else
+		return mcp25xxfd_set_raw_mode(dev,
+					      MCP25XXFD_OPMODE_NORMAL_CAN2);
+#endif
+	case CAN_LOOPBACK_MODE:
+		return mcp25xxfd_set_raw_mode(dev,
+					      MCP25XXFD_OPMODE_EXT_LOOPBACK);
+	default:
+		LOG_ERR("Unsupported CAN Mode %u", mode);
+	case CAN_SILENT_MODE:
+		return mcp25xxfd_set_raw_mode(dev,
+					      MCP25XXFD_OPMODE_LISTEN_ONLY);
+	}
+}
+
+static int mcp25xxfd_set_timing(const struct device *dev,
+				const struct can_timing *timing,
+				const struct can_timing *timing_data)
+{
+	struct mcp25xxfd_data *dev_data = DEV_DATA(dev);
+	union mcp25xxfd_nbtcfg nbtcfg;
+	union mcp25xxfd_dbtcfg ndtcfg;
+	union mcp25xxfd_tdc tdc;
+	union mcp25xxfd_tscon tscon;
+	uint8_t mode;
+	int ret;
+
+	if (!timing || !timing_data) {
+		return -EINVAL;
+	}
```

Comment:
For regular CAN frames, *timing_data is parsed as NULL.
You probably need an extra ifdef to check if canfd is enabled then.